### ISSUE TITLE
Image API

### ISSUE THREAD

**Issue by cryptoquick** [nannou-org/nannou#155]
I love how easy this is to use, but I don't see anything regarding images. How do you intend for that to work, and is there any way I can help, as a new-ish Rust developer?

**Comment by cryptoquick** [comment]
I apologize if I wasn't clear earlier.

My goal is to make an image gallery, and I really like the robustness of the file drop functionality, I had trouble implementing that with Relm / GTK.

I'd like to paint image rects in the UI, and also have a scrolling mechanism as well. I think I could make the scrolling mechanism, I just don't know how to paint an image.

I realize there's a comment on the README about images not being implemented. Let me know if there's anything I can do to help. I noticed there's access to a lower-level image painting API in one of the methods in the UI code, but I'm not sure if that's the appropriate place to start.

**Comment by JoshuaBatty** [comment]
Thanks for opening this issue @cryptoquick . We are working on putting together an Image API to use with upcoming Vulkan switch in nannou 0.9. I've sketched out a rough start for what the Image API could look like below. 

```
pub struct Image {}

impl Image {
    /// Construct a new image module with an initial size
    pub fn new() -> Self {
        Image {}
    }
    //Load an image from disk
    pub fn load(path: Path) {}
    
    //-------- These functions operate in place ----------//
    
    //Resizes the width and height of the loaded image
    //Effeectively scaling up or scaling down
    pub fn resize(w: f32, h: f32) {} 
    //Crops the image to the new bounds
    pub fn crop(x: f32, y: f32, w: f32, h: f32) {}
    //Flip an image horizontally
    pub fn flip_horizontal() {}
    //Flip an image vertically
    pub fn flip_vertical() {} 
    // Rotate an image 90 degrees clockwis
    pub fn rotate90() {} 
    //Rotate an image 180 degrees clockwise
    pub fn rotate180() {} 
    //Rotate an image 270 degrees clockwise 
    pub fn rotate270() {} 
    //Convert the supplied image to grayscale
    pub fn grayscale() {} 
    //Invert each pixel within the supplied image
    pub fn invert() {} 

    //-------- Drawing ----------//

    // This will act like the geometry draw, with builder methods 
    // for setting .x_y() .w_h() .rotate() .rgba() .hsv() etc...
    pub fn draw() -> Self {}
    // Alternative to draw(). Allows the user to define a subsection of the image
    // to be drawn. All common draw builder methods can still be used with this
    pub fn draw_subsection(sx: f32, sy: f32, sw: f32, sh: f32) -> Self {}

    //-------- Pixel Control ----------//
    // These are taken from image-rs crate. Maybe we could 
    // Wrap the Pixel type into our own Nannou Pixel type? 

    // Return the pixel located at (x, y)
    pub fn get_pixel(x: f32, y: f32) -> Self::Pixel;
    /// Put a pixel at location (x, y)
    pub fn put_pixel(x: f32, f: u32, pixel: Self::Pixel);
    /// Return an Iterator over the pixels of this image.
    /// The iterator yields the coordinates of each pixel
    /// along with their value
    pub fn pixels() -> Pixels<Self>;

    // Writes and returns the images pixels into a GPU Frame Buffer
    pub fn to_fbo() -> Fbo {}
    // Assumes an fbo has been created that holds the GPU 
    // texture of the image. Returns as reference to the FBO
    // that can be passed into shader targets for example. 
    pub fn fbo_texture() -> &Fbo {}

    //-------- Retreive Information about the image ----------//

    //Return the width of the image
    pub fn width() -> f32 {}
    //Return the height of the image
    pub fn height() -> f32{}
    //Return a tuple containing the width and height of the image
    pub fn dimensions() -> (f32, f32) {}
    /// The bounding rectangle of this image.
    pub fn bounds() -> (f32, f32, f32, f32);

    //Save the image to disk
    pub fn save(path: Path) {}
}

```

**Comment by cryptoquick** [comment]
Nice! I'm looking forward to it. I've since made an Electron app, but I'd prefer to port it to Rust once things develop.

**Comment by mitchmindtree** [comment]
OK, this has been implemented for the most part! #484 includes the majority of the work by enabling support for "textuers" in the `draw` API.  You can think of a `Texture` as an instance of an image that lives on the GPU. Check out the new `examples/draw/draw_texture*` examples for a demonstration of how to draw with textures in a variety of different ways. Also, be sure to checkout out the generative design `image` examples.

If you have come across this issue and are still missing something from the image API, or you are unsure of how to do something, feel free to open a new issue!
